### PR TITLE
core/assert: print backtrace on failed assertion

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

If the `backtrace` module is enabled, use it to print a backtrace on a failed assertion. 

### Testing procedure

Add

    USEMODULE += backtrace

(currently only implemented on `native`).

Now if an assertion fails:

```
main(): This is RIOT! (Version: 2022.07-devel-395-g2b9c0-backtrace-assert)
Hello World!
You are running RIOT on a(n) native board.
This board features a(n) native MCU.
examples/hello-world/main.c:27 => 0x80497c7
0x8049621
0x8049681
0x804980e
0x80497d8
*** RIOT kernel panic:
FAILED ASSERTION.

*** halted.
```

### CODE REVIEW

**Comment by fjmolinas** [comment]
Can you show some log output?

**Comment by benpicco** [comment]
Sure! I just added `assert(0)` to `examples/hello-world` and added the `backtrace` module.

Now unfortunately `backtrace` is only implemented on `native`, so this is not as useful as I had initially hoped.
But nothing stops us for adding a backtrace implementation for e.g. ARM too.

**Comment by chrysn** [comment]
The documentation in `assert.h` describes that `assert()`'s behavior depends on DEBUG_ASSERT_VERBOSE; it could be updated to also indicate the influence of the BACKTRACE module.

Will this do something sane when an assertion fails in an interrupt handler? (Someone explicitly using backtrace would probably check this themselves, but placing an assert is something innocuous).

**Comment by maribu** [comment]
> I think it should not make a difference, this is only unwinding stack frames, AFAIR there should be no difference if those are on the interrupt stack or the thread stack.

This is not generally true, as some platforms only back up the program counter and leave backing up the registers to software. E.g. on AVR this is the case each ISR can have a unique stack frame format. Even when you force GCC to spill all registers on to the stack in an ISR, GCC will back up registers in different order depending on its current mood.

@nandojve does have a solution to that, though: https://github.com/RIOT-OS/RIOT/pull/16211

**Comment by maribu** [comment]
What is needed to get this in?

**Comment by benpicco** [comment]
Just a review.

**Review by maribu — APPROVED** [review]
ACK. Looks good to me. But let's wait for the feature freeze before hitting merge.

There seem to be some test failing - maybe they were waiting for a specific assert message and got of the rails due to changed console output.

File: core/lib/assert.c
Comment (miri64) [inline core/lib/assert.c:31]: As there is no line feed in the line above, the alternative code path should add one (but keep the old format in check, if it is not included)

```suggestion
#if IS_USED(MODULE_BACKTRACE)
    printf("failed assertion. Backtrace:\n"
    backtrace_print();
#endif
```

**Review by miri64 — CHANGES_REQUESTED** [review]

**Review by maribu — APPROVED** [review]

**Review by miri64 — APPROVED** [review]
bors merge

**Comment by benpicco** [comment]
bors cancel

bors merge

**Comment by benpicco** [comment]
bors merge

**Comment by kaspar030** [comment]
bors cancel
bors merge

**Pull request merged**
